File: s20182019/hw6/hw6.py
import numpy as np
from numpy.linalg import inv

N = 61

phi = np.zeros(shape=(101,N),dtype=np.float64)
phi[:][:] = 0.33374

T = 300.0   # Temperature (K)
kB = 1.38066E-23    # Boltzmann constant (J/K)
N_acc = 1.0E+24  # Dopant density (# / cm^3)
Ni = 1.075E+16  # Intrinsic density
q = 1.602192E-19 # Elementary charge (C)

Vt = kB*T/q    # Thermal voltage, V

eps0 = 8.85E-12 # Vacuum Permitivity
eps1 = 11.7 # Si permitivity (relative)
eps2 = 3.9 # SiO2 permitivity (relative)

dx = 0.1e-9 # 0.1 nm Spacing

mid1 = 5;   # interface at x=0.5 nm
mid2 = 55;  # interface at x=5.5 nm
coef = dx*dx*q/eps0
phi_now = np.zeros(shape=(N),dtype=np.float64)

e_d = np.zeros(shape=(101,N),dtype=np.float64)
e_d_int = np.zeros(shape=(101),dtype=np.float64)

for j in range(101):
    if j > 0:
        phi_now = phi[j-1]
        phi_now[0] = phi_now[0] + 0.01*j
        phi_now[N-1] = phi_now[0]
    else:
        phi_now = phi[j]
    for i in range(20):
        J = np.zeros(shape=(N,N),dtype=np.float64)  # Jacobian
        res = np.zeros(shape=(N),dtype=np.float64)
        res[0] = phi_now[0] - (0.33374+0.01*j)
        J[0][0] = 1.0
        res[N-1] = phi_now[N-1] - (0.33374+0.01*j)
        J[N-1][N-1] = 1.0
        for ii in range(1,N-1): # Laplacian part
            if ii < mid1 or ii > mid2:
                res[ii] = eps2*(phi_now[ii+1]-2.0*phi_now[ii]+phi_now[ii-1])
                J[ii][ii-1] = eps2
                J[ii][ii] = -2.0*eps2
                J[ii][ii+1] = eps2
            elif ii is mid1:
                res[ii] = eps1*(phi_now[ii+1]-phi_now[ii])+eps2*(phi_now[ii-1]-phi_now[ii])
                J[ii][ii-1] = eps2
                J[ii][ii] = -eps1-eps2
                J[ii][ii+1] = eps1
            elif ii is mid2:
                res[ii] = eps1*(phi_now[ii-1]-phi_now[ii])+eps2*(phi_now[ii+1]-phi_now[ii])
                J[ii][ii-1]=eps1
                J[ii][ii] = -eps1-eps2
                J[ii][ii+1] = eps2
            else:
                res[ii] = eps1*(phi_now[ii+1]-2*phi_now[ii]+phi_now[ii-1])
                J[ii][ii-1] = eps1
                J[ii][ii] = -2*eps1
                J[ii][ii+1] = eps1
        for ii in range(mid1,mid2+1):
            if ii is mid1:
                res[ii] = res[ii] - coef*(N_acc+Ni*np.exp(phi_now[ii]/Vt))*0.5
                J[ii][ii] = J[ii][ii] - coef*Ni*np.exp(phi_now[ii]/Vt)/Vt*0.5
            elif ii is mid2:
                res[ii] = res[ii] - coef*(N_acc+Ni*np.exp(phi_now[ii]/Vt))*0.5
                J[ii][ii] = J[ii][ii] - coef*Ni*np.exp(phi_now[ii]/Vt)/Vt*0.5
            else:
                res[ii] = res[ii] - coef*(N_acc+Ni*np.exp(phi_now[ii]/Vt))
                J[ii][ii] = J[ii][ii] - coef*Ni*np.exp(phi_now[ii]/Vt)/Vt

        
        update = np.dot(inv(J),-res)
        phi_now = phi_now + update
    
    phi[j] = phi_now

    for i in range(mid1,mid2+1):
        e_d[j][i] = Ni * np.exp(phi_now[i]/Vt)
        if i is mid1 or i is mid2:
            e_d_int[j] += dx*0.5*e_d[j][i]
        else:
            e_d_int[j] += dx*e_d[j][i]

#    f = open("d"+str(j)+".dat",'w')
#    for k in range(N):
#        f.write("{}\t{}\n".format(k,phi_now[k]))

f = open("integrated_density.dat",'w')
for j in range(101):
    f.write("{}\t{}\n".format(0.01*j,e_d_int[j]))
f.close()

File: s20172106/HW_6.m
clear all;
clc

%% Defining variables

q = 1.602192e-19; % Elementary charge, C
eps0 = 8.854187817e-12; % Vacuum permittivity, F/m
k_B = 1.380662e-23; % Boltzmann constant, J/K
T = 300.0; % Temperature, K
thermal = k_B*T/q; % Thermal voltage, V
Deltax = 0.1e-9; % 0.1 nm spacing
N = 61; % 6 nm thick
x = Deltax*transpose([0:N-1]); % real space, m
interface1 = 6; % At x=0.5 nm
interface2 = 56; % At x=5.5 nm
eps_si = 11.7; eps_ox = 3.9; % Relative permittivity
Nacc = 1e24; % 1e18 /cm^3
ni = 1.075e16; % 1.075e10 /cm^3
coef = Deltax*Deltax*q/eps0;
Work_Ftn=-4.30; % vacuum level-fermi level [eV]
Int_Fermi=-4.63374; % vacuum level-intrinsic fermi level [eV]
Volt_Gate = [-1:0.1:1]'; % set: fermi level= 0 V at equilibrium 
kk = size(Volt_Gate); % The number of the different Volt_Gate's

phi = zeros(N,kk(1));
%% Potential by using the Newton method
Contact_Volt = zeros(kk(1),1);

for k = 1:kk(1) %% or kk(1) for Gate voltage from 0 V to 1 V 

Contact_Volt(k) = ((Work_Ftn - Volt_Gate(k,1))- Int_Fermi);

for newton = 1:1000
  

    res = zeros(N,1);
    Jaco = sparse(N,N);
    res(1,1) = phi(1,k) - Contact_Volt(k);
    Jaco(1,1) = 1.0;
    res(N,1) = phi(N,k) - Contact_Volt(k);
    Jaco(N,N) = 1.0;
    %%%%%% Laplacian part %%%%%%%%%%%%%%%%%%%%%%%%
    
    for ii = 2:N-1
        if (ii< interface1 || ii> interface2)
            res(ii,1) = eps_ox*phi(ii+1,k) - 2*eps_ox*phi(ii,k) + eps_ox*phi(ii-1,k);
            Jaco(ii,ii-1) = eps_ox; Jaco(ii,ii) = -2*eps_ox; Jaco(ii,ii+1) = eps_ox;
        elseif (ii == interface1)
            res(ii,1) = eps_si*phi(ii+1,k) - (eps_ox + eps_si)*phi(ii,k) + eps_ox*phi(ii-1,k);
            Jaco(ii,ii-1) = eps_ox; Jaco(ii,ii) = - (eps_ox + eps_si); Jaco(ii,ii+1) = eps_si;
        elseif (ii == interface2)
             res(ii,1) = eps_ox*phi(ii+1,k) - (eps_ox + eps_si)*phi(ii,k) + eps_si*phi(ii-1,k);
            Jaco(ii,ii-1) = eps_si; Jaco(ii,ii) = - (eps_ox + eps_si); Jaco(ii,ii+1) = eps_ox;
        else
             res(ii,1) = eps_si*phi(ii-1,k) - 2*eps_si*phi(ii,k) + eps_si*phi(ii+1,k);
            Jaco(ii,ii-1) = eps_si; Jaco(ii,ii) = -2*eps_si; Jaco(ii,ii+1) = eps_si;
        end
    end
    
    %%%%%%% If there are charges (==> Poisson's eqn) %%%%%%%%%%%%%%%%%%%%%%%%%%%%
    
    for ii = interface1:interface2
         if (ii == interface1)
             res(ii,1) = res(ii,1) - coef*(Nacc + ni*exp(phi(ii,k)/thermal))*0.5;
             Jaco(ii,ii) = Jaco(ii,ii) - coef*ni*exp(phi(ii,k)/thermal)/thermal*0.5;
         elseif (ii == interface2)
             res(ii,1) = res(ii,1) - coef*(Nacc + ni*exp(phi(ii,k)/thermal))*0.5;
             Jaco(ii,ii) = Jaco(ii,ii) - coef*ni*exp(phi(ii,k)/thermal)/thermal*0.5;
         else
             res(ii,1) = res(ii,1) - coef*(Nacc + ni*exp(phi(ii,k)/thermal));
             Jaco(ii,ii) = Jaco(ii,ii) - coef*ni*exp(phi(ii,k)/thermal)/thermal;
         end
    end
    
    update = Jaco \ (-res);
    phi(:,k) = phi(:,k) + update;
end

end       
            
%% The integrated electron density
elec = zeros(N,kk(1));
Inte_elec = zeros(kk(1),1);
for s = 1:kk(1)

for ii=interface1:interface2
    elec(ii,s)=ni*exp(phi(ii,s)/thermal);
end
Inte_elec(s) = trapz(elec(:,s))/1e8; % Scaling for the integrated electron density 1/cm^2

end
figure
for tt = 1:kk(1)
semilogy(x,elec(:,tt)); hold on
end
xlabel('Position [nm]');
ylabel('n_e [1/cm^3]]');
figure
semilogy(Volt_Gate,Inte_elec); 
xlabel('Gate Voltage [V]');
ylabel('Integrated n_e [1/cm^2]]');


